### PR TITLE
Fix gradient transforms with default qubit and overridden shot vectors

### PR DESCRIPTION
**Pull request by albi3ro**
[sc-44395]

With this change, we can now do:
```
dev = qml.device("default.qubit")

@qml.qnode(dev)
def circuit(x):
    qml.RX(x, wires=0)
    return qml.expval(qml.PauliZ(0))

x = qml.numpy.array(0.543, requires_grad=True)

qml.gradients.param_shift(circuit)(x, shots=(10000, 10000, 1000) )
```
```
(array(-0.5209), array(-0.5258), array(-0.522))
```

### CODE REVIEW

**Review by mudit2812 — APPROVED** [review]
Thanks Christina! I'm glad this is working now. Could you explain why this works? I don't really understand why.

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -48,10 +48,13 @@
 
 <h3>Bug fixes 🐛</h3>
 
+* Gradient transforms now work with overridden shot vectors and default qubit.
```
Comment (timmysilv) [inline doc/releases/changelog-dev.md:51]:
```suggestion
* Gradient transforms now work with overridden shot vectors and default qubit.
  [(#4795)](https://github.com/PennyLaneAI/pennylane/pull/4795)
```

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -48,10 +48,13 @@
...
+
 * `qml.defer_measurements` now correctly transforms circuits when terminal measurements include wires
   used in mid-circuit measurements.
   [(#4787)](https://github.com/PennyLaneAI/pennylane/pull/4787)
 
+
```
Comment (timmysilv) [inline doc/releases/changelog-dev.md:57]:
```suggestion
```

**Review by timmysilv — APPROVED** [review]
we were popping the shots! nice find

**Comment by albi3ro** [comment]
@mudit2812 @timmysilv More or less how I figured this out:


First I saw that shots weren't on the tape getting sent to `param_shift`.  So where was that tape getting created? 

Instead of getting created by the call to `QNode.construct` inside of `QNode.__call__`, it was being created by the call to `QNode.construct` inside `TransformProgram._set_all_classical_jacobians`.  But by the time the kwargs got to `_set_all_classical_jacobians`, shots had been popped out by the first call to `QNode.construct`.

If a bug is weird, it's usually in-place mutation causing problems in a separate part of the code base.
